# Теория вероятностей и математическая статистика
##  Урок 8. Дисперсионный анализ. Логистическая регрессия
### Задача№1

**1. Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов. Даны значения роста в трех группах случайно выбранных спортсменов:  
Футболисты: 173, 175, 180, 178, 177, 185, 183, 182.  
Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180.  
Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.**

In [1]:
import numpy as np
import pandas as pd
from scipy import stats

In [2]:
football=np.array([173, 175, 180, 178, 177, 185, 183, 182], np.float64)
print(f'рост футболистов {football}')
hockey=np.array([177, 179, 180, 188, 177, 172, 171, 184, 180], np.float64)
print(f'рост хоккеистов {hockey}')
weightlifting=np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170], np.float64)
print(f'рост штангистов {weightlifting}')

n1 = len(football)
n2 = len(hockey)
n3 = len(weightlifting)
n = n1 + n2 + n3
print(f'количество участвующих в анализе \nфутболистов - {n1}, \nхоккеистов - {n2}, \nштангистов - {n3}.\nвсего спортсменов - {n}')

рост футболистов [173. 175. 180. 178. 177. 185. 183. 182.]
рост хоккеистов [177. 179. 180. 188. 177. 172. 171. 184. 180.]
рост штангистов [172. 173. 169. 177. 166. 180. 178. 177. 172. 166. 170.]
количество участвующих в анализе 
футболистов - 8, 
хоккеистов - 9, 
штангистов - 11.
всего спортсменов - 28


In [3]:
football_mean = np.mean(football)
hockey_mean = np.mean(hockey)
weightlifting_mean = np.mean(weightlifting)
print(f'Проведем однофакторный дисперсионный анализ. Сначала найдем средние для \nфутболистов = {football_mean}, \nхоккеистов = {hockey_mean}, \nштангистов = {weightlifting_mean}')

Проведем однофакторный дисперсионный анализ. Сначала найдем средние для 
футболистов = 179.125, 
хоккеистов = 178.66666666666666, 
штангистов = 172.72727272727272


Видно, что средний рос разнится. Установим, что это отличие статистически значимо. Для этого сначала соберем все значения роста в один массив:

In [4]:
sport_all = np.concatenate([football, hockey, weightlifting])
sport_all

array([173., 175., 180., 178., 177., 185., 183., 182., 177., 179., 180.,
       188., 177., 172., 171., 184., 180., 172., 173., 169., 177., 166.,
       180., 178., 177., 172., 166., 170.])

In [5]:
sport_mean = np.mean(sport_all)
print(f'Найдем среднее значение роста по всем значениям: {sport_mean}')

Найдем среднее значение роста по всем значениям: 176.46428571428572


In [6]:
s2 = np.sum((sport_all - sport_mean)**2)
print(f'Найдем  S2  — сумму квадратов отклонений наблюдений от общего среднего: {s2}')

Найдем  S2  — сумму квадратов отклонений наблюдений от общего среднего: 830.9642857142854


In [7]:
s2_f = ((football_mean - sport_mean )**2) * n1 + ((hockey_mean - sport_mean )**2) * n2 + ((weightlifting_mean - sport_mean )**2) * n3
print(f'Найдем сумму квадратов отклонений средних групповых значений от общего среднего: {s2_f}')

Найдем сумму квадратов отклонений средних групповых значений от общего среднего: 253.9074675324678


In [8]:
s2_ost = ((football - football_mean) ** 2).sum() + ((hockey - hockey_mean) ** 2).sum() + ((weightlifting - weightlifting_mean) ** 2).sum()
print(f'Найдем состаточную сумму квадратов отклонений: {s2_ost}')

Найдем состаточную сумму квадратов отклонений: 577.0568181818182


Удостоверимся, что соблюдается равенство $S^2 = S_F^2 + S_{ост}^2$:

In [9]:
if np.round(s2, 6) != np.round((s2_f + s2_ost), 6):
    print(f'равенство не соблюдается')
else:
    print(f'равенство верно')
print(f'незначительную разницу \n{s2} \nне равно \n{s2_f + s2_ost} \nотбросим округлив до 6-ого знако после запятой')

равенство верно
незначительную разницу 
830.9642857142854 
не равно 
830.964285714286 
отбросим округлив до 6-ого знако после запятой


In [10]:
sigma2_general = s2 / (n - 1)
print(f'Найдем общую дисперсию = {sigma2_general}')
k=3
sigma2_f = s2_f / (k - 1)
print(f'Найдем факторную дисперсию = {sigma2_f}, \nгде k = {k} это количество участвующих групп')

Найдем общую дисперсию = 30.776455026455015
Найдем факторную дисперсию = 126.9537337662339, 
где k = 3 это количество участвующих групп


In [11]:
sigma2_residual = s2_ost / (n - k)
print(f'Найдем остаточную дисперсию = {sigma2_residual}')

Найдем остаточную дисперсию = 23.08227272727273


In [12]:
F_h = sigma2_f / sigma2_residual
print(f'Вычислим  FH = {F_h}')

Вычислим  FH = 5.500053450812598


Найдем значение $F_{крит}$ в таблице критических точек распределения Фишера-Снедекора для заданного уровня значимости $\alpha = 0.05$ и двух степеней свободы: 

$df_{межд} = k - 1 = 3 - 1 = 2$ и $df_{внутр} = n - k = 28 - 3 = 25$.

$$F_{крит}= 3.38 < FH =5.50$$
Видим, что значения отличаются сильно, поэтому заключаем, что отличие среднего роста футболистов, хокеистов и штангистов статистически значимы.
ПРоверим табличные значения вычислениями.

In [13]:
alpha = 0.05

F_crit = stats.f.ppf(1 - alpha, k - 1, n - k)
F_crit

3.3851899614491687

# <center> Курсовой проект
### <center> A/B-тестирование

AБ-тестирование — это инструмент, который помогает проверять гипотезы и принимать решения на основе полученных данных. Метод заключается в разделении аудитории на части демонстрации разных вариантов чего-либо, сборе данных и оценке полученной информации. О A/B тестировании говорят часто в разрезе тестирования сайтов, можно провести аналогию и с другими продуктами, начиная от любых товаров и заканчивая сбором статистических данных в научных работах.

Какие вопросы позволяют решить A/B-тесты:
1. проведение анализа признаков отказов
2. создание обратной связи для снижения отказов и рисков при изменении
3. оценка важность чего - либо
4. гарантия статистически значимые улучшения ,к примеру в подаче продукта
5. позволяют для незначительных выборок провести анализ роли изминений для улучшения продукта и повышения окупаемости
6. уменьшение затрат за счет масштабирования продуктов и решений.


ААБ тестирование
Для более достоверных результатов, испытание лучше проводить с контрольной выборкой. Для этого надо поделить аудиторию, не на две группы (A и Б), а на три (A, A и Б). С двумя группами (основная и контрольная) провести один тест ,для третьей — другой. Это выявит количество случайных ошибок.

Нулевая гипотеза (H0) — некое предположение относящееся к результатам теста.Нулевая гипотеза считается верной,до тех пор пока не доказано обратное.

Альтернативная гипотеза (HA) – утверждение, противоположное нулевой гипотезе. Выдвигается, но не проверяется.

Уровень значимости (α) – вероятность отвергнуть верную нулевую гипотезу.
Нулевая гипотеза всегда проверяется на определенном уровне значимости. Например, если мы проверяем нулевую гипотезу на уровне значимости 5%, это означает, что если мы будем
проводить аналогичные исследования 100 раз и проверять на основе имеющихся данных интересующую нас нулевую гипотезу, в 5 случаях из 100 мы отвергнем нулевую гипотезу, хотя
она будет верной.

Значение статистического критерия (формула зависит от конкретной гипотезы) вычисляются по выборочным данным и выясняется, попадает ли оно в критическую область. При проверке гипотез есть вероятность допустить ошибку: первого рода (верная гипотеза отклонена, α) или второго рода (неверная гипотеза принята, β).

### Задача№2
Компания не осуществляет инвестиционных вложений в ценные бумаги с дисперсией
годовой доходности более чем 0,04. Выборка из 52 наблюдений по активу А показала, что
выборочная дисперсия ее доходности равна 0,045.Выяснить, допустимы ли для данной
компании инвестиционные вложения в актив А на уровне значимости 0,05.
***
$\alpha = 0,05$
Нулевая гипотеза:
$H0: \sigma^2=\sigma_0^2=0.04$  
Альтернативная гипотеза:  
$H1: \sigma^2>0.04$  
Вычисляем наблюдаемое значение критерия:
$$X_н^2 = \frac{(n-1)s_x^2}{\sigma_0^2} = \frac{(52-1)0.045}{0.04} = 57.375$$

Верна гипотиза $H1$
$$X_(кр)^2 (\alpha, n-1)= (0.05, 51)=>  68,669$$
$$X_(кр)^2>X_н^2$$
$$ 68,669>57,375 $$
На уровне значимости 0,05 можно принять и нулевую гипотезу.

## Задача №3
Есть некое приложение, позволяющее получать некую информацию. Недавно была добавлена разработка, для улучшения приложения, но результаты A/B этому утверждению противоречат. Задача понять нуж но ли оставлять доработку.

1. Первое что необходимо учесть, А/В это статистическое исследование в котором возможны ошибки еще на этапе формулирования задач.
2. Возможно разработка изначально не решала проблемм приложения и пользователя. 
3. В тесте разработки могли быть неправильно учтены критерии. 
4. Разработка может быть с некоторой "поломкой" в коде и существенно ухудшать работу сервиса. 
5. Последнее необходимо правильно понимать свою целевую аудиторию и проводить тесты именно на ней.

Соответственно необходимо соблюдать все этапы тестирования
1. Определяем цель
2. Определяем метрики
3. Формулируем гипотезу
4. Настраиваем эксперимент
5. Проводим эксперимент
6. Анализируем результаты


А/В-тест — инструмент, который не дает однозначного ответа на вопрос «Какой вариант лучше?», а лишь позволяет снизить неопределенность на пути к поиску оптимальных решений. При его проведении детали важны на всех этапах подготовки, каждая неточность стоит ресурсов и может негативно повлиять на достоверность результатов. Надеюсь, эта статья была полезной для вас и поможет избежать ошибок при А/В-тестировании.
